### ISSUE TITLE
A few small suggestions

### ISSUE THREAD

**Issue by trufanov-nok** [JohnLangford/vowpal_wabbit#726]
I. I've played with --readable_model and was quite surprised to find a lot of `0.000000` feature values in its output. VW doesn't save features with 0 weight to regressor. In a moment I realized that they're not zeroes but just very small. `gd.cc::save_load_regressor()` uses simple `:%f \n` for sprintf to save weights and `%f` has precession 6 by default. So you can't see values smaller than 1e-6 - you'll get only leading zeroes in readable model. My suggestion is to use `%g` instead:
```
f	Decimal floating point, lowercase	392.65
F	Decimal floating point, uppercase	392.65
e	Scientific notation (mantissa/exponent), lowercase	3.9265e+2
E	Scientific notation (mantissa/exponent), uppercase	3.9265E+2
g	Use the shortest representation: %e or %f	392.65
G	Use the shortest representation: %E or %F	392.65
```
This results in following output:
```
...
8246:-2.64074e-10
8264:0.605609
8265:-0.495252
8266:-0.596221
8288:-1.95944e-12
8289:-1.63608e-12
...
```

There is a drawback - if someone parses readable model files automatically he might not be ready to parse scientific format of real values. Also `%g` shall be slower than `%f`.

II. I found error messages `malformed example !\n'|' , ':' , space or EOL expected after : xxx` quite useless bcs they display not the whole example but its piece up to current `reading_head`. And this piece is perfectly fine bcs the problem happened just after reading_head. For example for `-1 |f val1:1.0val2:2.0` (space between features is missing) you'll get  `malformed example !\n'|' , ':' , space or EOL expected after : -1 |f val1:1.0`. Thus you have to open dataset, find this example and check the rest of example line to find the problem. I changed this to simply print whole example line in error message in my fork. There may be other solutions too.

**Comment by trufanov-nok** [comment]
III
 I'm writing a reduction and found out that `parser.cc::set_done(all)` is seems to be standard way to manually stop learning and exit VW.  Surprisingly I realized that reduction's end_examples() callback function is never invoked in this case. 
There is a bug at [this line](https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/learner.cc#L55)
```
  if (all.early_terminate) //drain any extra examples from parser and call end_examples
      while ( all.early_terminate == false )
{
}
```
while loop body won't be ever called.

**Comment by trufanov-nok** [comment]
IV. At least OAA and ECT are not compatible with link functions. For example, ECT [here](https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/ect.cc#L206) calls `base.learn` which invokes `link()` over `pred.scalar`. So if `link==logistic` is used it always be > 0. And thus on next line ECT always chooses the right leaf of tree.

**Comment by JohnLangford** [comment]
Both I. and II. seems good.

-John

**Comment by JohnLangford** [comment]
I'm not sure what to do with this.  In the case of logistic, at least, 
there is a version of logistic (glf1) which should be compatible.  
Perhaps this should be checked? At the point in scorer.cc where the 
logistic function is set it should be easy to check to see if ECT is 
called.

Why is this incompatible with OAA though?

-John

**Comment by JohnLangford** [comment]
@trufanov-nok do you want to add a check for ECT compatibility with a link function?   I could go either way.  The default link function is clearly ok, and we don't have a general mechanism to force people to use sensible reductions.

**Comment by trufanov-nok** [comment]
Can't estimate if I'll have enough time to dig into this in a next few weeks. But I wouldn't leave it as is without a warning at least. So how about a warning for known unsupported reductions?

**Comment by trufanov-nok** [comment]
I've taken a look into link problem again. It seems that OAA **is** compatible with `--link logistic`. It just require `--loss_function logistic` in this case. It doesn't work good for any other loss with `--link logistic` on a test dataset ( the predictions become equal after a few passes thus class 1 is always chosen). When logistic loss function is used the results are as good as for `--link glf1` or `--link identity` with default loss function.  

It might be an effect of the fact that OAA uses scalar prediction for training after it was scaled by link function while setting {-1;1} [label to example](https://github.com/trufanov-nok/vowpal_wabbit/blob/master/vowpalwabbit/oaa.cc#L78). So even if -1.f was predicted  for incorrect class it will be learned as -1 label and scalar 0.26 and got penalty for that. 

So, I guess ECT is the only known "by design" problem.

**Comment by trufanov-nok** [comment]
I've have played with ECT. Parameters are:
`vowpalwabbit/vw -k --ect 10 --error 3 -c --passes 10 --invariant -d test/train-sets/multiclass --holdout_off`

It gives `average loss = 0.110000` for all links except logistic. And `average loss = 0.900000` for logistic (all current predictions are 8).
It seems it might be fixed by just changing `> 0` to `> 0.5` [here](https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/ect.cc#L192) for link logistic case. And by adjusting right won condition in train [here](https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/ect.cc#L233).  The result will be same: `average loss = 0.110000`. So perhaps I can just fix ECT rather than add warnings?
P.S. I used `bool won = (2*ec.pred.scalar-1) * simple_temp.label > 0;` for train.
